### PR TITLE
ingress: promote CRUD API tests for v1 to conformance

### PR DESCRIPTION
**Pull request by cmluciano**
**What type of PR is this?**
 /kind feature
/kind api-change

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
xref: kubernetes/enhancements#1453

**Special notes for your reviewer**:

Builds on top of https://github.com/kubernetes/kubernetes/pull/89778, only the second commit is new

**DO NOT MERGE UNTIL THE FOLLOWING HAVE MERGED & ARE STABLE**
- [x] IngressClass CRUD API tests #91830
- [x] Ingress CRUD API tests #91593 

Run history:
* [gce-cos-master](https://testgrid.k8s.io/sig-release-master-blocking#gce-cos-master-default&include-filter-by-regex=sig-network.*Ingress.*AP&width=5)
* [kind-master-parallel](https://testgrid.k8s.io/sig-release-master-blocking#kind-master-parallel&include-filter-by-regex=sig-network.*Ingress.*API&width=5)
* current API coverage of those tests against the beta API visible at https://apisnoop.cncf.io/latest/beta/networking

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

- [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/1453-ingress-api/README.md)

### CODE REVIEW

**Comment by liggitt** [comment]
looks like it needs a rebase on the merged v1beta1 e2e tests

**Comment by liggitt** [comment]
`pull-kubernetes-conformance-kind-ga-only-parallel` will probably be unhappy until the admission bits merge as well

File: test/e2e/network/ingress.go [hunk 1]
```diff
@@ -947,50 +947,53 @@ var _ = SIGDescribe("Ingress API", func() {
...
-		prefixPathType := networkingv1beta1.PathTypePrefix
+		ingVersion := "v1"
+		ingClient := f.ClientSet.NetworkingV1().Ingresses(ns)
+
+		prefixPathType := networkingv1.PathTypePrefix
+		serviceBackend := &networkingv1.IngressServiceBackend{
+			Name: "defaultbackend",
+			Port: networkingv1.ServiceBackendPort{
+				Name:   "",
+				Number: 80,
```
Comment (liggitt) [inline test/e2e/network/ingress.go:967]: to be equivalent, should be:
```suggestion
				Number: 8080,
```

File: test/e2e/network/ingress.go [hunk 2]
```diff
@@ -947,50 +947,53 @@ var _ = SIGDescribe("Ingress API", func() {
...
-		ingVersion := "v1beta1"
-		ingClient := f.ClientSet.NetworkingV1beta1().Ingresses(ns)
-
-		prefixPathType := networkingv1beta1.PathTypePrefix
+		ingVersion := "v1"
+		ingClient := f.ClientSet.NetworkingV1().Ingresses(ns)
+
+		prefixPathType := networkingv1.PathTypePrefix
+		serviceBackend := &networkingv1.IngressServiceBackend{
+			Name: "defaultbackend",
```
Comment (liggitt) [inline test/e2e/network/ingress.go:964]:
```suggestion
			Name: "default-backend",
```

File: test/e2e/network/ingress.go [hunk 3]
```diff
@@ -947,50 +947,53 @@ var _ = SIGDescribe("Ingress API", func() {
...
+						IngressRuleValue: networkingv1.IngressRuleValue{
+							HTTP: &networkingv1.HTTPIngressRuleValue{
+								Paths: []networkingv1.HTTPIngressPath{{
 									Path:     "/",
 									PathType: &prefixPathType,
-									Backend: networkingv1beta1.IngressBackend{
-										ServiceName: "test-backend",
-										ServicePort: intstr.FromInt(8080),
-									},
+									Backend:  defaultBackend,
```
Comment (liggitt) [inline test/e2e/network/ingress.go:989]: keep this as equivalent to the existing test as possible, e.g. use `test-backend`, not defaultBackend

File: test/e2e/network/ingressclass.go [hunk 4]
```diff
@@ -115,25 +116,30 @@ func createIngressClass(cs clientset.Interface, name string, isDefault bool, uni
...
-		Spec: networkingv1beta1.IngressSpec{
-			Backend: &networkingv1beta1.IngressBackend{
-				ServiceName: "default-backend",
-				ServicePort: intstr.FromInt(80),
+		Spec: networkingv1.IngressSpec{
+			DefaultBackend: &networkingv1.IngressBackend{
+				Service: &networkingv1.IngressServiceBackend{
+					Name: "defaultbackend",
+					Port: networkingv1.ServiceBackendPort{
+						Name:   "",
```
Comment (liggitt) [inline test/e2e/network/ingressclass.go:132]:
```suggestion
```

File: test/e2e/network/ingressclass.go [hunk 5]
```diff
@@ -42,7 +43,7 @@ var _ = SIGDescribe("IngressClass [Feature:Ingress]", func() {
 		cs = f.ClientSet
```
Comment (liggitt) [inline test/e2e/network/ingressclass.go:43]: conformance tests probably should not have a `[Feature:Ingress]` tag... that prevents it from running in default e2e configs

File: test/e2e/network/ingressclass.go [hunk 6]
```diff
@@ -42,7 +43,7 @@ var _ = SIGDescribe("IngressClass [Feature:Ingress]", func() {
...
 	})
 
-	ginkgo.It("should set default value on new IngressClass", func() {
+	framework.ConformanceIt("should set default value on new IngressClass", func() {
 		ingressClass1, err := createIngressClass(cs, "ingressclass1", true, f.UniqueName)
```
Comment (liggitt) [inline test/e2e/network/ingressclass.go:47]: this assumes there is no default ingressclass present in the cluster already... not sure it is valid for an e2e to make assumptions about the existing cluster config, and almost certainly not valid for a conformance test to do so... that would mean a cluster that installs its own default ingress class could not be conformant, which doesn't seem right

File: test/e2e/network/ingressclass.go [hunk 7]
```diff
@@ -57,7 +58,7 @@ var _ = SIGDescribe("IngressClass [Feature:Ingress]", func() {
 		}
 	})
 
-	ginkgo.It("should not set default value if no default IngressClass", func() {
+	framework.ConformanceIt("should not set default value if no default IngressClass", func() {
 		ingressClass1, err := createIngressClass(cs, "ingressclass1", false, f.UniqueName)
```
Comment (liggitt) [inline test/e2e/network/ingressclass.go:62]: this test also assumes the cluster did not set up its own default ingress class... this doesn't seem correct to promote to conformance as-is

File: test/e2e/network/ingressclass.go [hunk 8]
```diff
@@ -70,7 +71,7 @@ var _ = SIGDescribe("IngressClass [Feature:Ingress]", func() {
 		}
 	})
 
-	ginkgo.It("should prevent Ingress creation if more than 1 IngressClass marked as default", func() {
+	framework.ConformanceIt("should prevent Ingress creation if more than 1 IngressClass marked as default", func() {
 		ingressClass1, err := createIngressClass(cs, "ingressclass1", true, f.UniqueName)
```
Comment (liggitt) [inline test/e2e/network/ingressclass.go:75]: this test is disruptive to an existing cluster... and can't be run in parallel with the other test in this suite... this needs to be resolved before promotion to conformance

File: test/e2e/network/ingressclass.go (hunk 6 above)
Comment (liggitt) [reply test/e2e/network/ingressclass.go:47]: also, this test can't be run in parallel with the other test in this suite... this needs to be resolved before promotion to conformance

File: test/e2e/network/ingressclass.go (hunk 7 above)
Comment (liggitt) [inline test/e2e/network/ingressclass.go:62]: also, this test can't be run in parallel with the other test in this suite... this needs to be resolved before promotion to conformance

File: test/e2e/network/ingressclass.go (hunk 5 above)
Comment (liggitt) [reply test/e2e/network/ingressclass.go:43]: open a separate PR to:
1. remove the feature tag
2. mark the tests that create a default ingressclass or which depend on there being no default ingressclass as `[Serial]`

Comment (liggitt) [reply test/e2e/network/ingressclass.go:43]: limit this PR to switching to use v1 types and promoting the API CRUD tests

Comment (cmluciano) [reply test/e2e/network/ingressclass.go]: opened issue to track #92147

**Comment by liggitt** [comment]
/lgtm
/approve
from an API perspective

/hold for v1 PRs to merge and for @kubernetes/cncf-conformance-wg review

**Comment by liggitt** [comment]
last commit promoting beta ingress/ingressclass API tests to conformance needs review

**Comment by liggitt** [comment]
https://github.com/kubernetes/kubernetes/pull/89778 is in the merge queue.

Once it merges, can rebase this, cancel hold, and get conformance approval

**Comment by liggitt** [comment]
Ready for rebase and hold cancel

**Comment by johnbelamaric** [comment]
Ok, run history looks good, and these changes are just to use the new API. Current guidance in the conformance test docs says these should "only promote" but we probably should update that to allow API tweaks to avoid our chicken & egg problem (though as my son says the egg *obviously* came first because it way laid by a proto-chicken).

/approve

**Pull request merged**